### PR TITLE
Makes SequenceReducer a LudwigModule, adds reduction_modules_test.py

### PR DESCRIPTION
**Pull request by dantreiman**
- Also adds doc strings to torch_utils.py
- Adds SequenceReducer max_sequence_length parameter
- Adds SequenceReducer embedding_size optional parameter

### CODE REVIEW

File: ludwig/modules/reduction_modules.py [hunk 1]
```diff
@@ -35,6 +35,22 @@ def __init__(self, reduce_mode=None, **kwargs):
     def forward(self, inputs, mask=None):
         return self._reduce_obj(inputs, mask=mask)
 
+    def infer_output_shape(self, input_shape):
```
Comment (ShreyaR) [inline ludwig/modules/reduction_modules.py:38]: This method seems to be doing a lot of what the `output_shape` property in a `LudwigModule` does.

You could make the parent class of `SequenceReducer` a `LudwigModule`, and add `input_shape` and `output_shape` properties. The `output_shape` would have the same logic that's implemented in `infer_output_shape`, and the `input_shape` can return a dummy tensor.

Comment (ShreyaR) [reply ludwig/modules/reduction_modules.py:38]: Additionally, can you update parts of the code that use `SequenceReducer` and then manually infer the shape of the output after reduction to use the `output_shape` property of the `SequenceReducer` instead?

E.g. here's some [code](https://github.com/ludwig-ai/ludwig/blob/389c1aa443ddc0b5be92981aa9811421d58491bd/ludwig/encoders/sequence_encoders.py#L793) that can use `SequenceReducer.output_shape`.

File: ludwig/utils/torch_utils.py [hunk 2]
```diff
@@ -15,27 +15,48 @@
 
 
 def sequence_length_2D(sequence: torch.Tensor) -> torch.Tensor:
-    """Returns the number of non-zero elements per sequence."""
+    """Returns the number of non-zero elements per sequence in batch.
```
Comment (ShreyaR) [inline ludwig/utils/torch_utils.py:18]: Thanks for adding this!

File: tests/ludwig/modules/test_reduction_modules.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.parametrize("reduce_mode", ["last", "sum", "mean", "avg", "max", "concat", None])
+@pytest.mark.parametrize("test_input_shape", [(16, 1, 4), (4, 10, 16)])
+def test_sequence_reducer(reduce_mode: str, test_input_shape: Tuple[int, ...]):
+    batch_size = test_input_shape[0]
+    max_sequence_length = test_input_shape[1]
+    sequence_reducer = reduction_modules.SequenceReducer(reduce_mode=reduce_mode)
+    inputs = torch.zeros(test_input_shape)
```
Comment (ShreyaR) [inline tests/ludwig/modules/test_reduction_modules.py:18]: Can you make sure these tests are GPU compliant? Involves moving the input tensor and module to GPU if available.

E.g. [here](https://github.com/ludwig-ai/ludwig/blob/1daab8714463e63c8f4b4ba4caed3cc1ef97a195/tests/ludwig/encoders/test_h3_encoders.py)

**Review by ShreyaR — COMMENTED** [review]
Thanks for the PR! Added comments.

File: ludwig/modules/reduction_modules.py (hunk 1 above)
Comment (dantreiman) [reply ludwig/modules/reduction_modules.py:38]: Haha, I missed LudwigModule._compute_output_shape.   However, the sequence reducer does not know the sequence length ahead of time, so I can’t fully specify output shape having input shape passed in somehow.

**Review by justinxzhao — APPROVED** [review]
LGTM! Using defaults for the additional args required by `FeedForwardAttentionReducer` seems reasonable.

**Review by ShreyaR — APPROVED** [review]
Looks great!

**Pull request merged**